File: typescript-node/src/rsa.ts
import getPrimeNumber from './prime-number';

export function encrypt(
    message: string,
    publicKey: bigint,
    exponent: bigint
): bigint[] {
    const encrypted: bigint[] = [];

    for (let i = 0; i < message.length; i++) {
        encrypted.push(
            bigIntModPow(BigInt(message.charCodeAt(i)), exponent, publicKey)
        );
    }

    return encrypted;
}

export function decrypt(
    message: bigint[],
    publicKey: bigint,
    privateKey: bigint
): bigint[] {
    const decrypted: bigint[] = [];

    for (const character of message) {
        decrypted.push(bigIntModPow(character, privateKey, publicKey));
    }

    return decrypted;
}

function bigIntModPow(
    number: bigint,
    exponent: bigint,
    modulus: bigint
): bigint {
    let result = 1n;

    for (let i = 0n; i < exponent; i++) {
        result = result * number;
        result = result % modulus;
    }

    return result;
}

export function calculateKeys(
    p: bigint,
    q: bigint
): {
    publicKey: bigint;
    publicExponent: bigint;
    privateKey: bigint;
} {
    const n = p * q;
    const f = (p - 1n) * (q - 1n);

    let e = getPrimeNumber(f.toString(2).length);
    let gcd = 0n;
    let yFactor: bigint;

    do {
        e += 1n;
        ({ gcd, yFactor } = calculateGreatestCommonDivisor(f, e));
    } while (gcd !== 1n);

    return {
        publicKey: n,
        publicExponent: e,
        privateKey: yFactor,
    };
}

function calculateGreatestCommonDivisor(
    _x: bigint,
    _y: bigint
): { gcd: bigint; yFactor: bigint } {
    let x = _x;
    let y = _y;

    let factor: bigint;
    let leftover: bigint;

    let z: bigint;
    let xx = 1n;
    let yy = 0n;
    let v = 1n;
    let u = 0n;

    while (y !== 0n) {
        factor = x / y;
        leftover = x - factor * y;

        // Setup for next iteration
        x = y;
        y = leftover;

        // Calculate Y-factor
        z = u;
        u = xx - factor * u;
        xx = z;
        z = v;
        v = yy - factor * v;
        yy = z;
    }

    return { gcd: x, yFactor: yy < 0 ? _x + yy : yy };
}

export function displayAsText(encrypted: bigint[]): string {
    return encrypted
        .map(e => String.fromCharCode(parseInt(e.toString())))
        .join('');
}

File: rust/src/rsa.rs
use num_bigint::{BigInt, ToBigInt};
use num_traits::{one, zero};

use crate::prime_number;

#[derive(Debug)]
pub struct Keys {
    pub public_key: BigInt,
    pub public_exponent: BigInt,
    pub private_key: BigInt,
}

pub fn encrypt(message: &str, public_key: &BigInt, exponent: &BigInt) -> Vec<BigInt> {
    message
        .chars()
        .map(|character| (character as u32).to_bigint().unwrap())
        .map(|num| num.modpow(&exponent, &public_key))
        .collect()
}

pub fn decrypt(message: &Vec<BigInt>, public_key: &BigInt, private_key: &BigInt) -> Vec<BigInt> {
    message
        .iter()
        .map(|character| character.modpow(&private_key, &public_key))
        .collect()
}

pub fn calculate_keys(p: &BigInt, q: &BigInt) -> Keys {
    let n = p * q;
    let f: BigInt = (p - 1) * (q - 1);

    let mut e = prime_number::get_prime_number(f.bits());

    let mut gcd: BigInt;
    let mut y_factor: BigInt;

    loop {
        e += 1;

        let output = calculate_greatest_common_divisor(&f, &e);
        gcd = output.0;
        y_factor = output.1;

        if gcd == one() {
            break;
        }
    }

    Keys {
        public_key: n,
        public_exponent: e,
        private_key: y_factor,
    }
}

fn calculate_greatest_common_divisor(_x: &BigInt, _y: &BigInt) -> (BigInt, BigInt) {
    let mut x = _x.clone();
    let mut y = _y.clone();

    let mut factor: BigInt;
    let mut leftover: BigInt;

    let mut z: BigInt;
    let mut xx: BigInt = one();
    let mut yy: BigInt = zero();
    let mut v: BigInt = one();
    let mut u: BigInt = zero();

    loop {
        factor = &x / &y;
        leftover = &x - &factor * &y;

        // Setup for next iteration
        x = y;
        y = leftover;

        // Calculate Y-factor
        z = u.clone();
        u = &xx - &factor * &u;
        xx = z;
        z = v.clone();
        v = &yy - &factor * &v;
        yy = z;

        if y == zero() {
            break;
        }
    }

    return (x, if yy < zero() { _x + yy } else { yy });
}

pub fn display_as_text(numeric: &Vec<BigInt>) -> String {
    let mut textual = String::new();

    for numeric_char in numeric {
        for digit in numeric_char.to_bytes_be().1 {
            textual.push(digit as char)
        }
    }

    return textual.escape_default().collect();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn calculates_gcd() {
        let (gcd, _) =
            calculate_greatest_common_divisor(&27.to_bigint().unwrap(), &18.to_bigint().unwrap());

        assert_eq!(gcd, 9.to_bigint().unwrap());
    }
}
